src/serac/numerics/functional/boundary_integral_kernels.hpp

```diff
@@ -109,237 +62,86 @@ template <int i>
-                   CPUArrayView<derivatives_type, 2> qf_derivatives, const mfem::Vector& J, const mfem::Vector& X,
-                   const mfem::Vector& N, std::size_t num_elements, lambda qf)
-      : qf_derivatives_(qf_derivatives), J_(J), X_(X), N_(N), num_elements_(num_elements), qf_(qf)
-  {
-  }
+    // (batch) evalute the q-function at each quadrature point
+    auto qf_outputs = batch_apply_qf(qf, x_e, J_e, get<indices>(qf_inputs)...);
 
-  /**
-   * @brief integrate the q-function over the specified domain, at the specified trial space values
-   *
-   * @param U input E-vectors
-   * @param R output E-vector
-   */
-  void operator()(const std::vector<const mfem::Vector*> U, mfem::Vector& R)
-  {
-    // mfem provides this information in 1D arrays, so we reshape it
-    // into strided multidimensional arrays before using
-    constexpr int sdim = dimension_of(geom) + 1;  // spatial dimension
-    constexpr int nqp  = num_quadrature_points<geom, Q>();
-    auto          J    = reinterpret_cast<const tensor<double, nqp>*>(J_.Read());
-    auto          X    = reinterpret_cast<const tensor<double, sdim, nqp>*>(X_.Read());
-    auto          N    = reinterpret_cast<const tensor<double, sdim, nqp>*>(N_.Read());
-    auto          r    = reinterpret_cast<typename test_element::dof_type*>(R.ReadWrite());
-    static constexpr TensorProductQuadratureRule<Q> rule{};
-
-    tuple u_e = {
-        reinterpret_cast<const typename decltype(type<indices>(trial_elements))::dof_type*>(U[indices]->Read())...};
-
-    // for each element in the domain
-    for (uint32_t e = 0; e < num_elements_; e++) {
-      // load the jacobians, positions and normals for each quadrature point in this element
-      auto J_e = J[e];
-      auto X_e = X[e];
-      auto N_e = N[e];
-
-      // batch-calculate values / derivatives of each trial space, at each quadrature point
-      tuple qf_inputs = {promote_each_to_dual_when<indices == differentiation_index>(
-          type<indices>(trial_elements).interpolate(get<indices>(u_e)[e], rule))...};
-
-      // (batch) evalute the q-function at each quadrature point
-      auto qf_outputs = batch_apply_qf(qf_, X_e, N_e, qf_inputs, Iseq);
-
-      qf_outputs = elementwise_multiply(qf_outputs, J_e);
-
-      // write out the q-function derivatives after multiplying by J_e so that
-      // won't need to be applied in the action_of_gradient and element_gradient kernels
+    // write out the q-function derivatives after applying the
+    // physical_to_parent transformation, so that those transformations
+    // won't need to be applied in the action_of_gradient and element_gradient kernels
+    if constexpr (differentiation_index != serac::NO_DIFFERENTIATION) {
       for (int q = 0; q < leading_dimension(qf_outputs); q++) {
-        qf_derivatives_(e, q) = get_gradient(qf_outputs[q]);
+        qf_derivatives[e * qpts_per_elem + uint32_t(q)] = get_gradient(qf_outputs[q]);
```

Comment:
I can't think of a way of handling this other than implicit conversions (obvious no) or wrapping the types to convert internally.  Does this bleed into the user code or is it just us?

Reply:
This affects any code that mixes some combination of containers from mfem/axom/stl. 

I don't believe there's any way to prevent implicit conversion between integer types, because that's part of the C++ language specification. Serac uses `-Wsign-conversion` which warns on such conversions, and we also use `-Werror` which makes those warnings into errors, so we're forced to cast everything back and forth. 

The ideal solution is to use containers that follow a consistent convention, so the casting is unnecessary and there's no implicit conversion, but that requires cooperation from mfem and axom. I think axom's index types are configurable, so we could ask it to use stl-compatible integer types, but mfem hardcodes `int` for everything.

related mfem issue: https://github.com/mfem/mfem/issues/3229

Reply:
I vote we make an issue and discuss at a later date. It is more important to get this PR in.